### ISSUE TITLE
Data retention CLI

### ISSUE THREAD

**Issue by DaimonPl** [AbsaOSS/spline#684]
Is there any easy way to configure automatic retention for spline data? For example automatically delete entires older than 3 months ?

**Comment by wajda** [comment]
No, not yet.  It could be added in the future though, it was slightly mentioned a couple of times in some local discussions.

**Comment by DaimonPl** [comment]
Ok, I think it might be useful in big setups. We have hundreds of jobs running everyday and dozens of jobs running every hour pretty sure we would need to scale arangodb quite often without retention

Any idea if retention can be somehow added in arangodb itself ?

**Comment by wajda** [comment]
Well...not in ArangoDB natively (afaik), but you could use your custom script. However this approach is kinda fragile and dirty, and I wouldn't recommend doing it. Especially in a busy environment.
To do it correctly and avoiding breaking the database consistency you need to respect the graph structure and cascade the deletion properly.

On the other hand it you could find a logical time gap in your data pipelines, big enough  (e.g. one hour), you can try to cut the DB in the middle or even closer to the end of that time gap. Simply delete all documents with `_created < my_timestamp` from all collections excepts `dataSource`, all in one transaction.
This is still a hacking approach but having the idle time frame large enough it practically should not create any problems.
Also keep in mind that the described approach might not work in Spline 0.6+

**Comment by DaimonPl** [comment]
Thx for detailed answer :)

No way to find any time gap on our Hadoop it's processing 24/7/365 something all the time :)

**Comment by wajda** [comment]
By the time gap I mean such time frame, for which there is no document in the Spline database with `from < _created < to`

**Comment by wajda** [comment]
Awesome. I see you case will be a really great test case for Spline :))

**Comment by pratapmmmec** [comment]
Thanks for responding. Do we have any tentative timeline for its release?

**Comment by wajda** [comment]
We'll try to implement something simple in 0.5.4 which should be out in the course of the month. We are addressing performance bottlenecks in this release, so this feature logically fits here.

**Comment by wajda** [comment]
Solution:
1. Add a command to the Admin CLI to remove the lineage data before older than given date.
2. Add an option to the Gateway to do the same thing periodically based on a provided cron expression.

**Comment by wajda** [comment]
After the second thought I realized that having CLI command for removing old data makes it unnecessary to make it a Spline Server responsibility. System cronjobs can be used instead to simply run `java -jar admin-cli.jar db-prune ...`

**Comment by wajda** [comment]
The issue is blocked by arangodb/arangodb#12693

**Comment by wajda** [comment]
Moving it to 0.6.0 as it most likely needs to wait until the ArangoDB issue is resolved.

**Comment by pratapmmmec** [comment]
@wajda  Do we have any workaround as of now? 
Our lineage is now growing at ~ 2GB on daily basis and its getting difficult to sustain at this rate. Ours is a 24 X 7 workload running all days. We are fine if the lineage older than 20 days is broken and to handle it we will can limit on pagination query to go back only for 20 days at max.

**Comment by wajda** [comment]
Yes, the blocking ArangoDb issue has been resilved and we'll return to this ticket soon.

**Comment by Aditya-Sood** [comment]
hi @wajda 

We are working to create a foxx service which will take care of Spline's arango collections cleanup, and the pseudocode for the same is as follows:

1. Based on a threshold value (say in days), remove all 'progress' documents (& 'progressOf' edges) older than the threshold days

2. Identify all orphan execution plans (to which no 'progressOf' edge points to)
	a. Delete all collections and edges belonging to the the orphan execution plans
	b. Delete all orphan execution plans

3. Identify all orphan data sources (to which no 'affects' or 'depends' edge points to)
	a. Delete all such orphan data sources

We will update you on how our testing goes and if any challenges are discovered during the same

Kindly let us know if you have any other requirements for us to consider

Thank you!

cc: @pratapmmmec @vishalag001

**Comment by wajda** [comment]
@Aditya-Sood , thank you for your efforts! Yes that is roughly how it should work. There is however a few pitfalls that you need to be aware of:
1. > Identify all orphan data sources. 

Remember that the DataSource has a _many-to-many_ relationship with the Read/Write operations and are inserted _before_ the operations that use them, so there is a risk to delete a data source that has just been created and the operation connected to it has yet to be inserted. Basically we could be potentially dealing with the partially committed transactions and low isolation level, which unfortunately could both occur in the ArangoDB cluster setup. If such a rare case occurs you could leave the database in a logically inconsistent state as there is no foreign key constraints in ArangoDB and it _not_ stop you from adding an edge pointing to a non-existing node. So be careful with removing data source nodes.

2. > Identify all orphan execution plans

The same as above but with respect to the execution event -> execution plan relationship. Depends on how you use Spline, especially if there are standing jobs somewhere executed once in awhile that generates the same execution plan (with the same hash based UUID) you might need to account for late events.

3. In the upcoming Spline 1.0.0 GA we are planing to support application level transactions to remediate for lack of full ACID (especially A and I) in the ArangoDB cluster. To accomplish that we will introduce a certain protocol that every producer and consumer must follow to guarantee proper Tx isolation. This will involve a special metadata associated with every node and edge in the database. So the script that you write will either be not working after upgrade, or it will be working but would not be properly isolated, making the above mentioned problem more likely as we are abandoning Arango native Tx support and will only be relying on our own application level transactions.

**Comment by wajda** [comment]
Here's an old branch where I started adding data retention commands to the Spline Admin CLI - https://github.com/AbsaOSS/spline/tree/feature/spline-684-data-retention
Unfortunately due to a critical bug that was discovered in the ArangoDB during working on that functionality the work was suspended. The bug has been later resolved, so I believe we will soon resume the work on the feature. This is just FYI. But of course any help and feedback is more than welcome! :)

**Comment by Aditya-Sood** [comment]
hi @wajda 
Thank you for highlighting the issue with lack of transaction isolation!

As a workaround we have decided to extend the threshold filter from the progress events to both the executionPlan and dataSource collections as well - this should allow the cleanup script to ignore any pseudo-orphan objects that have been created recently, and pick only those objects which are at least thresholdDays old

Kindly let us know if you foresee any other concerns with the same

Also the js program is now ready and we will be testing it on our arango db this week, will update you post the same

Thanks

cc: @pratapmmmec @vishalag001

**Comment by wajda** [comment]
Awesome! Thanks a million.

**Comment by Aditya-Sood** [comment]
hi @wajda 
We developed a straightforward script to iterate over all the collections, filter out relevant documents and then remove them; something like as follows:
```
db._query('FOR collectionEle IN @@cols FILTER collectionEle._created < @purgeLimitTimestamp FOR orphanExecPlanID IN @arr FILTER collectionEle._belongsTo == orphanExecPlanID REMOVE collectionEle IN @@cols',  
        {
            '@cols': collections[i],
            'arr': orphanExecPlanIDsOlderThanThresholdDaysArray,
            'purgeLimitTimestamp': purgeLimitTimestamp
        }
    )
```
However its execution turned out to be very slow, with the arangosh timing out during the processing of large collections such as `operation`

We tried restricting the purge window to just 1 days worth of data, but even then it timed out on the `uses` collection

This could be explained by the fact that this approach is primarily dependent on using the `_belongsTo` attribute for identifying the relevant documents, and since this attribute is not indexed for most collections the script ends up going through all the documents in it

As a workaround we are working on a graph traversal based approach to identify the `_key`s of all the edges and vertices linked to the orphan execution plans (since it is an indexed attribute), and then use that for removing the documents

Will update once we have had a chance to test the new approach
Hopefully the graph traversal will offer the necessary performance gain required for the deletion to go through without arangosh timing out

Thanks!

cc: @pratapmmmec @vishalag001

**Comment by wajda** [comment]
Did you try to create indices for `_belongTo` in all collections? Usually traversals are much slower than document style queries, especially on indexed attributes. That's why `_belongsTo` was created.

**Comment by Aditya-Sood** [comment]
Oh I was under the impression that traversals would be much faster since all the relevant collections (vertices + edges) belonging to a particular plan are being picked up together in one go?

Thanks for clarifying, will re-try the original document queries script after creating an index on `_belongsTo` and update

**Comment by wajda** [comment]
> Oh I was under the impression that traversals would be much faster since all the relevant collections (vertices + edges) belonging to a particular plan are being picked up together in one go?

No, unfortunately it doesn't work like that. Traversals basically go node by node conditionally jumping over edges. They are good for querying _graphs_ not because they are faster, but because you don't have any other options and just have to travel through edges to find nodes you are looking for.

**Comment by Aditya-Sood** [comment]
hi @wajda 
adding indexes on `_belongsTo` on the the remaining collections did the trick, we were able to get all the queries to run within the default 20 min server timeout threshold

however the subsequent query to identify orphan dataSources seems to need much more time than that (despite the query attributes being indexed), so we are tweaking the timeout threshold to see if it can be made to complete in a reasonable amount of time
will update post that

**Comment by wajda** [comment]
Just out of my curiosity, what's the size of your collections?

**Comment by hugeshi** [comment]
Hi @wajda ,
Thanks for your contribution to this issue, I have kept an eye on this feature for a long time. It's quite tough to clean up data on a big dataset. I think it's the main reason why several people request the data retention feature.
what's the rough ETA of this feature?

**Comment by wajda** [comment]
Unfortunately I cannot give any ETA. On one side, it's not a priority feature for our company at the moment. But, I can say that a lot of things are changing right now, and there is a chance for the project to receive some boost very soon. On the other side, as I tried to explain above this feature depends on another feature that introduces a set of fundamental changed to the way how transactions are managed, and that one has to be finished first. I hope to give some update on plans say next month.

**Comment by Aditya-Sood** [comment]
hi @wajda 
Apologies for the delay in response, we were caught up in some other tasks

Our total spline database size was around ~100GB on the system that we were trying to cleanup

We dealt with the `dataSource` objects deletion by taking two steps:
1. Trim down the `affects` and `depends` collections down to the target duration that we finally want, before initiating search for stale data sources (this reduced the overall dataset size that had to be traversed to identify the stale sources)
2. Rewrite the query being used to identify stale `dataSource` objects - there was an optimisation opportunity since using an equality operation on the non-indexed `_id` attribute was performing better than a string match on the indexed `_key` attribute

Combining this with the previous script changes have allowed us to trim a day's worth of stale objects in 1-1.5 hrs

**Comment by Aditya-Sood** [comment]
Following is the final cleanup script that we are using: https://gist.github.com/Aditya-Sood/ecc07c9f296dbdf03d4946c5d1b4efce

Could you please review the script and let us know if there are any changes that you would like?

Post that we would like to discuss integrating the cleanup functionality with the project, possibly as a configurable service added to the Admin utility

**Comment by wajda** [comment]
Thank you @Aditya-Sood. I will review it this week and will get back to you.

**Comment by wajda** [comment]
@Aditya-Sood, we've reviewed your script and found it generally correct. But I've got a few minor questions/suggestions about it:
1. Stage 1:
    - `FILTER progEvent.timestamp` vs. `FILTER progEvent._created` in the `FOR progEvent IN progress`. The difference is that `timestamp` denotes the _event_ time, while `_created` denotes the _processing_ time. Although having a large enough time window there is little to no practical difference, but I guess it logically it would be more correct to use the former.
    - Using `_created` on `progressOf`. Again, while it might be highly unlikely but still there is a risk of grabbing an edge that belongs to a living node just because its processing time happened to be 1ms earlier than the adjacent _progress_ node. Which could happen as there is no predetermined order of both entities creation and the `_created` time might not be exactly the same for all nodes inserted by a single logical transaction. This is the same reason why we can't just wipe all the docs from all collections in one go by the `_created_` predicate alone. I would suggest remove `progressOf` by looking at its relation to the `progress` being removed. Use either traversal `1 OUTBOUND prog progressOf` or a simple document query `FOR progOf IN progressOf FILTER progOf._from == prog._id` that could potentially be faster for the case.
2. At the stage 2 where you collect orphan execution plan IDs:
    - Why do you need a second loop. Wouldn't it better to collect them at the stage 1 when removing _progress_ and _progressOf_ nodes?
    - When checking for node orphanage, why do you use `NOT IN` and require two-step process? Can't you simply count the number of adjacent edges and compare it with `0`?
3. What do you mean by _"using an equality operation on the non-indexed `_id` attribute was performing better than a string match on the indexed `_key` attribute"_? In ArangoDB attributes `_id`, `_key`, `_from` and `_to` are always indexed.

**Comment by Aditya-Sood** [comment]
hi @wajda 
thanks for the review! please excuse the delay in response, I was busy with work and then Diwali over the past few days

I've understood the changes required for the first 3 subpoints
Regarding sub-point 4 (`When checking for node orphanage,...`), if I'm unable to use your suggestion from sub-point 3 (i.e. collect orphan exec plans in Stage 1 itself), then I'll add this as well

regarding point (3), the `dataSource` does not have an index on the `_id` attribute in our DB:
<img width="106" alt="Screenshot 2022-10-26 at 9 28 10 AM" src="https://user-images.githubusercontent.com/15438998/197931246-78378021-709b-4c3a-957d-f8947424efcf.png">
also I think the major difference came in because a sub-string search on an indexed attribute will not benefit from indexing, since all values will have to be scanned by such an operation

lastly, shall I continue working on this feature or has it been re-assigned?
I was under the impression this requirement was on the back-burner and not urgent, but if that has changed and you would prefer having @dk1844 continue do let us know

thanks

**Comment by wajda** [comment]
> the dataSource does not have an index on the _id attribute in our DB:

There is no need for a separate index for `_id`.  In ArangoDB, `_id` (also called a _document handle_) is just a combination of a collection name and a primary key (`_key`). If you take a look at the explain output, you'll see that the same primary index is used for a selection by the `_id` or `_key` attribute.

**Comment by wajda** [comment]
> lastly, shall I continue working on this feature or has it been re-assigned?
I was under the impression this requirement was on the back-burner and not urgent, but if that has changed and you would prefer having @dk1844 continue do let us know

As we didn't hear from you for some time we decided to speed things up a bit. We took your gist and (with some minor changes) basically incorporated it into our older PR that has been started and abandoned some long time ago. @Aditya-Sood it would be awesome if you help us testing and optimize @dk1844's PR trying it on your large database.

**Comment by dk1844** [comment]
In reference to https://github.com/AbsaOSS/spline/issues/684#issuecomment-1277297133, I did some measurement runs for stage 1 and stage 2 with different AQL to get a sense how they perform.

## Common info
All measuring was done on the same 12200 lineage records generated by our data-gen using
```bash for ($i = 0; $i -lt 1220; $i++){ echo "loop $i" ; docker run -e SPLINE_URL=http://host.docker.internal:8080/spline -e OPERATIONS=1-10@1 spline_datagen}```

Arango profiling was always used, done on "my machine", so the results should only be interpretted as relative to each other.
Also, timestamp filtering was commented out each time.

## Stage 1
### Option A
This option come from the [original PR#762 on this very topic](https://github.com/AbsaOSS/spline/pull/762/files#diff-1da771da430502f37edcb8b73dc7a9cd8bf939123e75e121a99c4ae50fd87ae8R27).
```aql
FOR p IN progress
    //FILTER p.timestamp < ${timestamp}
    REMOVE p IN progress
    FOR po IN progressOf
        FILTER po._from == p._id
        REMOVE po IN progressOf
        RETURN DISTINCT po._to
```				
### Option B
This option comes from the [@Aditya-Sood's script](https://gist.github.com/Aditya-Sood/ecc07c9f296dbdf03d4946c5d1b4efce). It has the theoretical downside of possibly creating orphanated progress and progressOf entries (that would get cleaned up on the next purge run)
```aql
FOR progEvent IN progress
    // FILTER progEvent._created < ${purgeLimitTimestamp}
    REMOVE progEvent IN progress

FOR progOf IN progressOf
    // FILTER progOf._created < ${purgeLimitTimestamp}
    REMOVE progOf IN progressOf
```
	
### Option C
This options comes from the reference suggestion message.
```aql
FOR p IN progress
    //FILTER p.timestamp < ${timestamp}
    FOR prog, progOf IN 1 OUTBOUND p progressOf
        REMOVE p IN progress
        REMOVE progOf IN progressOf
```

### Results of profiling
All results have max index selectivity (100%) on one or multiple indices.

_Option A_ - Peak Mem 1376256B; median time 2.20633s (2.17843s, 2.20633s, 2.42307s)
_Option B_ - Peak Mem   98304B; median time 1.51415s (1.53055s, 1.51415s, 1.53688s)
_Option C_ - Peak Mem 1540096B - median time 2.48856s (1.82244s, 2.26623s, 2.47635s, 2.48856s, 2.60309s, 2.29054s, 3.11524s)

### Possible conclusion
@Aditya-Sood's approach is the fastest and most memory conservative (_Option B_). If we don't want to use it because of possible inconsistency issue, other two options are relatively similar with _Option A_ seeming a bit better,

## Stage 2
Stage 2 has been attempted on the same data after one of the Stage 1 options has been run (regardless of which).
### Option A
This is @Aditya-Sood's approach with the two commands passed together (possibly faster).
```aql
    LET execPlanIDsWithProgressEventsArray = (FOR execPlan IN executionPlan
    FOR progOf IN progressOf
        FILTER progOf._to == execPlan._id // && execPlan._created < ${timestamp}
        COLLECT ids = execPlan._id
        RETURN ids)

    FOR execPlan IN executionPlan
        FILTER execPlan._id NOT IN execPlanIDsWithProgressEventsArray //&& execPlan._created < ${timestamp}
        RETURN execPlan._id
```
### Option B
This options comes from the reference suggestion message.
```aql
    FOR execPlan IN executionPlan
        LET connectionCount = LENGTH(FOR progOf IN progressOf
            FILTER progOf._to == execPlan._id // && execPlan._created < ${timestamp}
            RETURN 1
        )
        FILTER connectionCount == 0
        RETURN execPlan._id
````
### Results of profiling
All results have max index selectivity (100%) on one or multiple indices.

_Option A_ - Peak Mem 327680B; median first-time run 0.70598s, median of subsequent runs (cached?): 0.03214s 
_Option B_ - Peak Mem 950272B; median first-time run 0.75554s, median of subsequent runs (cached?): 0.04231s

### Possible conclusion
_Option A_ seems a bit better (less than half of memory needed, run time seems slightly better)
